docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+
+**Signaling the ban to p2p**
+
+As it is the mempool **reactor** that has access to the p2p layer, not the actual mempool implementation, the peer banning function will most likely have to send the peer ID to a channel to inform the reactor of the fact that this peer should be banned. This would require adding a channel for communication into the `CListMempool` on construction, and adding a routine in the mempool ractor that waits on a response via this channel. However, the actual implementation of this is yet to be defined.
+
+**Implementing peer banning on recheck**
+
+Currently the recheck logic confirmes whether once **valid** transactions, 
+, where `ResponseCheckTx.code == 0`, are still valid. 
+
+As this logic loops through the transactions in any case, we can leverage it to check whether we can ban peers. 
+However, this approach has several downsides:
+- It is not timely enough. Recheck is executed after a block is committed, leaving room for a bad peer to send 
+us transactions the entire time between two blocks. 
+- If we want to keep track of when peers sent us a traansaction and punish them only if the misbehaviour happens 
+frequently enough, this approach makes it hard to keep track of when exactly was a transaction submitted. 
+- Rechecking if optional and node operators can disable it. 
+
+On the plus side this would avoid adding new logic to the mempool caching mechanism and keeping additional information about
+transaction validity. But we would still have to keep the information on peers and the frequency at which they send us bad transactions.
+
+Transactions that became invalid on recheck should not be cause for peer banning as they have not been gossiped as invalid transactions. 
+
+#### `PreCheck` and`PostCheck`
+
+The `PreCheck` and `PostCheck` functions are optional functions that can be executed before or after `CheckTx`. 
+Following the design outlined in this RFC, their responses are not considered for banning. 
+
+
+#### Checks outside `CheckTx`
+
+There are a number of checks that the mempool performs on the transaction, that are not part of `CheckTx` itself. 
+Those checks, have been mentioned in the user issues described at the beginning of this document:
+
+- Transaction size
+- Proto checks
+
+The previous code snippets do not incroporate these in peer banning. If we adopt those as valid reasons for banning, we should put the corresponding logic in place. 
+
+### Impacted mempool functionalities
+
+- Mempool caching: remembering failed transactions and whether they come from banned peers; Removal of transactions from 
+  `invalidCachedTx` when a transaction is removed from cache.
+- Handling of transactions failing `CheckTx`: Keeping track of how often transactions from a particular peer have failed
+  and banning them if the conditions for a ban are met.
+
+### Impact on ABCI 
+
+- Introduction of new response codes for CheckTx
+- Altering the specification to reflect this change
+
```

Comment:
Should we also consider using bogus transactions (those which could never be valid) as evidence of misbehavior?